### ISSUE TITLE
Diameter unlocks vs techtree

### ISSUE THREAD

**Issue by Jimbodiah** [shadowmage45/SSTULabs#705]
Every new career I run into the issue that the maximum diameters for the parts are always too small for the place in my career. i.e. I am launching 3.75m rockets with boosters when I am going to Mars and by the time I unlock 6.25m I am already at the point in the career where I am using ion and NF propulsion parts and no longer need 6.25m+ lifters anymore.

Would you be open to moving each unlock down one node in the tree? This would mean unlokcing 1.875m at the third node instead of the fourth and unlock 10m tanks at the last node (instead of 6.25m). This would go for all part diameters.

**Comment by shadowmage45** [comment]
The default configurations should mimic stock.  So wherever stock unlocks a diameter, is where SSTU should unlock it.

Where are the 1.875m parts unlocked in the MH expansion in career mode?

5m+ parts are a bit more difficult to place, as stock barely goes to 3.75m.  On that subject, there -used- to be a 10m unlock in the last node ('experimentalsomething'), but that got removed with the change over to using CTT support.  (e.g. install CTT, and you'll see the 10m unlock)

**Comment by Jimbodiah** [comment]
No clue to the 1.875 unlock in 1.4/MH: My favorite mod is not compatible yet for 1.4 ;)   I spent like 30 minutes on 1.4 before I ran away screaming.

Yeah, this comes back to the never-ending discussion that the stock tree sucks balls. I've actually thought about making my own tree as even in CTT things are anything but balanced as each modmaker has his own interpretation as where to put stuff:  A warp drive is unlocked half way for one mod, F1 engines in the end on other mods ;)    

A further complication is that I tend to run RSS, which just needs bigger engines and tanks faster than a stock system would. This is why stock doesn't really do much over 2.5m as you barely need it to be honest, maybe a 3.75m core for really heavy missions.

I've already moved some parts around in the stock tree, but it's a PITA in the end anyway. I really can't find the energy to make my own tree or mod an existing one to make sense, but I guess I will have to at some point :)

**Comment by Jimbodiah** [comment]
Have you seen this one?

https://spacedock.info/mod/855/Historical%20Progression%20Tech%20Tree

It has linear branches in different areas (engines, aero, tanks etc) and each new node is a fixed era.

**Comment by Jimbodiah** [comment]
Perhaps it's an option to do this?

1.875m -> unchanged

2.5m -> unchanged

3.125m   ->  3.75m

3.75m -> 5m  (or 6.25m)

6.25m -> 10m


Not sure if adding the NEEDS[RSS] is an option for upgrades?  I'm trying this right now in a patch.

**Comment by shadowmage45** [comment]
> Not sure if adding the NEEDS[RSS] is an option for upgrades?  

Not in SSTU patches.  Any RO/RSS stuff should be done in the RO patches for SSTU.

> 3.125m -> 3.75m  

Only viable if I am for some reason not unlocking 3.75m at the same point as stock?

Please keep in mind that the default configuration is intended for 'stock-compatible' balancing -- so the diameters really should not unlock before they would with stock parts.

**Comment by Jimbodiah** [comment]
Who uses stock parts???  ;)

**Comment by taterkerman** [comment]
I am messing with a 1.5 career right now, and I have 3.75 stock heat shield unlocked, and the max SSTU I have is 1.25.

**Comment by taterkerman** [comment]
In career, all the SSTU heat shields >1.25m unlocked after I unlocked "Heavy Landing." In the 300 column. The 3.75m stock one unlocks at a 160 point node in the tree.

**Comment by taterkerman** [comment]
To be clear, if the goal is to mirror stock, it's not working. I never even saw 2.5m.

**Comment by shadowmage45** [comment]
What do the upgrades in the parts say?  

Or... I guess let me rephrase that... what exactly is the problem?   (what specifically needs to be moved, from where, and to where?)

Should be a fairly simple affair to fix it up once I know what is going on...

**Comment by taterkerman** [comment]
Sorry. The Modular Heat Shield is not upgrading at all until you unlock the Heavy Landing node (300 points). I unlocked (and purchased) the entry level version, then kept unlocking tree nodes (and buying upgrades where offered). No change in size available until that node that unlocks the 10m inflatable heat shield, then all the sizes unlocked. I can do more specific testing if you like---I can start a career with maxed funds and science, then serially unlock stuff til I see heat shield upgrades appearing.

**Comment by shadowmage45** [comment]
I was referring more to links to wherever this data is, such as -->  https://github.com/shadowmage45/SSTULabs/blob/master/GameData/SSTU/Parts/Upgrades/UpgradeParts-MHS.cfg

2.5m unlocks with 'landing' technology -- https://github.com/shadowmage45/SSTULabs/blob/master/GameData/SSTU/Parts/Upgrades/UpgradeParts-MHS.cfg#L16
3.75m unlocks with 'advLanding' technology -- https://github.com/shadowmage45/SSTULabs/blob/master/GameData/SSTU/Parts/Upgrades/UpgradeParts-MHS.cfg#L25
10m unlocks with 'heavyLanding' technology -- https://github.com/shadowmage45/SSTULabs/blob/master/GameData/SSTU/Parts/Upgrades/UpgradeParts-MHS.cfg#L34

How do those tech-nodes line up with the stock tech nodes?  I haven't touched career mode in... quite some time... so do not remember where the stock ones are located.

**Comment by shadowmage45** [comment]
Looking at the stock heat shields, they unlock at the same nodes as the SSTU ones -- 
* 2.5m unlocks with 'landing'  
* 3.75m unlocks with 'advLanding'

Could you confirm that these unlocks are -not- being respected for the SSTU parts in career mode?  (start a new career game, cheat in a bunch of science, unlock up to the respective node, go into VAB and test heat shield max diameter).  The part-upgrades that unlock these sizes should also show up in the tech-tree (if they are not present, that would point to other issues), and also note that you have to _purchase_ these upgrades for them to be available/applied.

There could be some issues/bugs with the plugin code, but the configs for the upgrades appear to be correct.  

---

Edit:  Config for the part itself also appears to be correct, and these are all using the _stock_ portion of the upgrade system whereby it is supposed to adjust the KSPField's of the PartModule automatically (my plugin code does not touch anything regarding upgrades in the MHS).

Is it possible that you have upgrades/unlocks disabled in the stock adv. settings menu?

**Comment by taterkerman** [comment]
I have them enabled (Part Upgrades is green).

Made a new career just now (sorry, got home, and was doing electrical work to rewire a sconce).

**Comment by taterkerman** [comment]
Survivability unlocks the 1.25m shield.

OK, this is weird. In my "real" career (that caused this post), I have to pay funds, not just science. I cannot make that an option in a new career. 

To be clear, this is an entirely NEW 1.5.1 career. Nothing left over from old versions, 100% clean install.

**Comment by taterkerman** [comment]
1.4.3 also no longer allows that difficulty setting where you pay science AND funds for career mode. Why did one career save enable this nonexistent option on a clean install?

I think it must be related to this, but I literally cannot see how to toggle that setting.

My persistent file has:
			BypassEntryPurchaseAfterResearch = False

The test careers have this True, and everything works (SSTU wise regarding upgrades). I did not notice the False setting, since that's how I always used to play career.

**Comment by shadowmage45** [comment]
Interesting, thanks for the confirmation that I wasn't losing my mind :)  (configs and plugin looked correct, figured there was a different reason for this to be not working).

I'll try and take a look this evening to see if I can track down what is going on with the save game settings.  IIRC it is a setting when you first create the game, related to something about entry-purchase-costs; I'm not sure if that setting can be altered after the game has been created.

**Comment by taterkerman** [comment]
What's weird is that it is not even an option in 1.5.1 that I can see. I never update my old version, I download the full installer, and each KSP version I use is entirely clean to start. Sometimes I later drag craft files into saves, but past that, nothing moves between versions for me.

**Comment by atomontage04** [comment]
I can confirm that latest version has a bug with Modular Heat Shield upgrades. Currently they are not visible in the tech tree.

I've filed a PR that fixes that.
https://github.com/shadowmage45/SSTULabs/pull/827
